macros/src/main/scala/mainecoon/autoMonad.scala

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2018 Jentsch
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mainecoon
+
+import scala.annotation.{StaticAnnotation, compileTimeOnly}
+import scala.meta._
+import Util._
+
+import collection.immutable.Seq
+
+/**
+  * auto generates an instance of `cats.Functor`
+  */
+@compileTimeOnly("Cannot expand @autoMonad")
+class autoMonad extends StaticAnnotation {
+  inline def apply(defn: Any): Any = meta {
+    enrichAlgebra(defn, higherKinded = false)(autoMonad.monadInst)
+  }
+}
+
+
+object autoMonad {
+  private[mainecoon] def monadInst(ad: AlgDefn): TypeDefinition = {
+    import ad._
+    import cls._
+
+    val flatMapMethods = templ.stats.toList.flatMap(_.collect {
+      case q"def $methodName[..$mTParams](...$params): ${Type.Name(`effectTypeName`)}" =>
+        q"""def $methodName[..$mTParams](...$params): TTarget =
+           mapFunction(delegatee_.$methodName(...${argumentss(params)})).$methodName(...${argumentss(params)})"""
+    })
+
+    val pureMethods = templ.stats.toList.flatMap(_.collect {
+      case q"def $methodName[..$mTParams](...$params): ${Type.Name(`effectTypeName`)}" =>
+        q"""def $methodName[..$mTParams](...$params): TTarget =
+           pureValue"""
+    })
+
+    val instanceDef = Seq(q"""
+      implicit def ${Term.Name("monadFor" + name.value)}[..$extraTParams]: _root_.cats.Monad[$typeLambdaVaryingEffect] =
+        new _root_.cats.Monad[$typeLambdaVaryingEffect] {
+          override def flatMap[T, TTarget](delegatee_ : $name[..${tArgs("T")}])(mapFunction: T => $name[..${tArgs("TTarget")}]): $name[..${tArgs("TTarget")}] =
+            new ${Ctor.Ref.Name(name.value)}[..${tArgs("TTarget")}] {
+              ..$flatMapMethods
+            }
+
+          override def pure[TTarget](pureValue: TTarget): $name[..${tArgs("TTarget")}] =
+            new ${Ctor.Ref.Name(name.value)}[..${tArgs("TTarget")}] {
+              ..$pureMethods
+            }
+
+          override def tailRecM[A, B](a: A)(f: A => $name[..${tArgs(t"Either[A, B]")}]): $name[..${tArgs("B")}] =
```

Comment:
I am curious if you've looked at the [monad implementation of `Function` in cats](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/instances/function.scala#L68-L76)? That would make it stack-safe.

Reply:
I tried, but my first attempt was fruitless. But as soon, as I understood the idea and the types, it was quite easy. 96ddd76 fixes the issue.